src/test/moment/start_end_of.js

```diff
@@ -393,3 +393,27 @@ test('endOf millisecond and no-arg', function (assert) {
     assert.equal(+m, +m.clone().endOf('millisecond'), 'endOf with millisecond argument should change time');
     assert.equal(+m, +m.clone().endOf('milliseconds'), 'endOf with milliseconds argument should change time');
 });
+
+test('startOf for year zero', function (assert) {
+    var m = moment('0000-02-29T12:34:56.789Z').parseZone();
+    assert.equal(m.clone().startOf('ms').toISOString(), '0000-02-29T12:34:56.789Z',  'startOf millisecond should preserver year');
+    assert.equal(m.clone().startOf('s').toISOString(), '0000-02-29T12:34:56.000Z', 'startOf second should preserver year');
+    assert.equal(m.clone().startOf('m').toISOString(), '0000-02-29T12:34:00.000Z', 'startOf minute should preserver year');
+    assert.equal(m.clone().startOf('h').toISOString(), '0000-02-29T12:00:00.000Z', 'startOf hour should preserver year');
+    assert.equal(m.clone().startOf('d').toISOString(), '0000-02-29T00:00:00.000Z', 'startOf day should preserver year');
+    assert.equal(m.clone().startOf('M').toISOString(), '0000-02-01T00:00:00.000Z', 'startOf month should preserver year');
+    assert.equal(m.clone().startOf('Q').toISOString(), '0000-01-01T00:00:00.000Z', 'startOf quarter should preserver year');
+    assert.equal(m.clone().startOf('y').toISOString(), '0000-01-01T00:00:00.000Z', 'startOf year should preserver year');
+});
+
+test('endOf for year zero', function (assert) {
+    var m = moment('0000-02-29T12:34:56.789Z').parseZone();
+    assert.equal(m.clone().endOf('ms').toISOString(), '0000-02-29T12:34:56.789Z',  'endOf millisecond should preserver year');
+    assert.equal(m.clone().endOf('s').toISOString(), '0000-02-29T12:34:56.999Z', 'endOf second should preserver year');
+    assert.equal(m.clone().endOf('m').toISOString(), '0000-02-29T12:34:59.999Z', 'endOf minute should preserver year');
+    assert.equal(m.clone().endOf('h').toISOString(), '0000-02-29T12:59:59.999Z', 'endOf hour should preserver year');
+    assert.equal(m.clone().endOf('d').toISOString(), '0000-02-29T23:59:59.999Z', 'endOf day should preserver year');
+    assert.equal(m.clone().endOf('M').toISOString(), '0000-02-29T23:59:59.999Z', 'endOf month should preserver year');
+    assert.equal(m.clone().endOf('Q').toISOString(), '0000-03-31T23:59:59.999Z', 'endOf quarter should preserver year');
+    assert.equal(m.clone().endOf('y').toISOString(), '0000-12-31T23:59:59.999Z', 'endOf year should preserver year');
+});
```

Comment:
Please add some more tests about the issues you are referencing.

For example, I'm not convinced this PR will solve https://github.com/moment/moment/issues/1990
There are examples in https://github.com/moment/moment/issues/2749, https://github.com/moment/moment/issues/3580,  and in https://github.com/moment/moment/issues/3132, https://github.com/moment/moment/issues/4152.

We should be able to construct test in particular locales. I'm not sure about testing in conjunction with moment-tz, but let me know how far you get.

Reply:
I'm not sure how to add tests that'll run on Travis for some of these things...

(Issue #952 addresses the difficulty of this and talks about a brute-force approach running tests across all timezones using a `grunt zones` task - but that task doesn't exist...)

When I was working on tests, I'd try and identify a specific date and timezone showing the problem, and run eye-ball a test run from a bash script.  Something like this:
```bash
#!/bin/bash
moment=./build/umd/moment.js

echo "Issue 2749:"
TZ='Europe/Copenhagen' node -e "$(cat << EOF
var moment = require('${moment}');
console.log("Expected:", moment('2015-10-25T02:00:00.000+01:00').unix());
console.log("Actual:  ", moment('2015-10-25T02:00:00.000+01:00').startOf('hour').unix());
EOF
)"

echo "Issue 3132:"
TZ='America/Sao_Paulo' node -e "$(cat << EOF
var moment = require('${moment}');
// Bug: Brazil was showing wrong end-of-day for 2016-10-16:
console.log("Expected: 2016-10-16 23:59:59");
console.log("Actual:   " + moment("2016-10-16").endOf('day').format("YYYY-MM-DD HH:mm:ss"));

// Check start of day:
console.log("Expected: 2016-10-16 01:00:00");
console.log("Actual:   " + moment("2016-10-16T12:34").startOf('day').format("YYYY-MM-DD HH:mm:ss"));
EOF
)"
``` 

Talking about #1990 specifically though: that one is fixed because the problem is in browser variation in how setMinutes/Seconds/Milliseconds is implemented (e.g. https://bugzilla.mozilla.org/show_bug.cgi?id=1084547)  - by avoiding these methods, we avoid the bug.